communication/src/AsyncTlsConnection.cpp

```diff
@@ -0,0 +1,390 @@
+  EC_KEY_free(ecdh);
+
+  // Only allow using the strongest cipher suites.
+  SSL_CTX_set_cipher_list(ssl_context_.native_handle(), tlsTcpImpl_.config_.cipherSuite.c_str());
+}
+
+bool AsyncTlsConnection::verifyCertificateClient(bool preverified,
+                                                 boost::asio::ssl::verify_context& ctx,
+                                                 NodeNum expected_dest_id) {
+  std::string subject(256, 0);
+  X509* cert = X509_STORE_CTX_get_current_cert(ctx.native_handle());
+  if (!cert) {
+    LOG_ERROR(logger_, "No certificate from server at node " << expected_dest_id);
+    return false;
+  }
+  X509_NAME_oneline(X509_get_subject_name(cert), subject.data(), 256);
+  auto [valid, _] = checkCertificate(cert, "server", subject, expected_dest_id);
+  (void)_;  // unused variable hack
+  return valid;
+}
+
+bool AsyncTlsConnection::verifyCertificateServer(bool preverified, boost::asio::ssl::verify_context& ctx) {
+  std::string subject(512, 0);
+  X509* cert = X509_STORE_CTX_get_current_cert(ctx.native_handle());
+  if (!cert) {
+    LOG_ERROR(logger_, "No certificate from client");
+    return false;
+  }
+  X509_NAME_oneline(X509_get_subject_name(cert), subject.data(), 512);
+  auto [valid, peer_id] = checkCertificate(cert, "client", std::string(subject), std::nullopt);
+  peer_id_ = peer_id;
+  return valid;
+}
+
+std::pair<bool, NodeNum> AsyncTlsConnection::checkCertificate(X509* receivedCert,
+                                                              std::string connectionType,
+                                                              std::string subject,
+                                                              std::optional<NodeNum> expectedPeerId) {
+  // First, perform a basic sanity test, in order to eliminate a disk read if the certificate is
+  // unknown.
+  //
+  // The certificate must have a node id, as we put it in `OU` field on creation.
+  //
+  // Since we use pinning we must know who the remote peer is.
+  // `peerIdPrefixLength` stands for the length of 'OU=' substring
+  int peerIdPrefixLength = 3;
+  std::regex r("OU=\\d*", std::regex_constants::icase);
+  std::smatch sm;
+  regex_search(subject, sm, r);
+  if (sm.length() <= peerIdPrefixLength) {
+    LOG_ERROR(logger_, "OU not found or empty: " << subject);
+    return std::make_pair(false, 0);
+  }
+
+  auto remPeer = sm.str().substr(peerIdPrefixLength, sm.str().length() - peerIdPrefixLength);
+  if (0 == remPeer.length()) {
+    LOG_ERROR(logger_, "OU empty " << subject);
+    return std::make_pair(false, 0);
+  }
+
+  NodeNum remotePeerId;
+  try {
+    remotePeerId = stoul(remPeer, nullptr);
+  } catch (const std::invalid_argument& ia) {
+    LOG_ERROR(logger_, "cannot convert OU, " << subject << ", " << ia.what());
+    return std::make_pair(false, 0);
+  } catch (const std::out_of_range& e) {
+    LOG_ERROR(logger_, "cannot convert OU, " << subject << ", " << e.what());
```

Comment:
Not worth factoring this out for both exceptions, so, duplication it is. Thanks C++.